File: Problems/No.144_Binary_Tree_Preorder_Traversal/binary_tree_preorder_traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        
        # parameters
        ans = []
        temp = root
        
        # start compute, using Morris Traversal
        while temp:
            if (temp.left == None):
                ans.append(temp.val)
                temp = temp.right
            else:
                prev = temp.left
                while (prev.right and prev.right!=temp):
                    prev = prev.right
                if (prev.right == None):
                    ans.append(temp.val)
                    prev.right = temp
                    temp = temp.left
                else:
                    prev.right = None
                    temp = temp.right
        
        return ans

File: Problems/No.144_Binary_Tree_Preorder_Traversal/binary_tree_preorder_traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        
        // parameters
        vector<int> ans;
        TreeNode *temp = root, *prev;
        
        // start compute, using Morris Traversal
        while (temp) {
            
            if (temp->left == NULL) {
                
                ans.push_back(temp->val);
                temp = temp->right;
                
            }
            else {
                
                prev = temp->left;
                while (prev->right and prev->right!=temp) prev = prev->right;
                if (prev->right == NULL) {
                    
                    ans.push_back(temp->val);
                    prev->right = temp;
                    temp = temp->left;
                    
                }
                else {
                    
                    prev->right = NULL;
                    temp = temp->right;
                    
                }
                
            }
            
        }
        
        return ans;
        
    }
};
